filebeat/reader/line/line.go

```diff
@@ -52,143 +64,81 @@ func New(input io.Reader, codec encoding.Encoding, bufferSize int) (*Reader, err
+		n, err := l.reader.Read(buf)
+		if err != nil {
+			return 0, err
 		}
+		l.buffer.Append(buf[:n])
 
-		if buf[len(buf)-1] == '\n' {
-			break
-		} else {
-			logp.Debug("line", "Line ending char found which wasn't one: %s", buf[len(buf)-1])
-		}
+		idx = l.buffer.Index(l.nl)
 	}
 
-	// output buffer contains complete line ending with '\n'. Extract
-	// byte slice from buffer and reset output buffer.
-	bytes, err := r.outBuffer.Collect(r.outBuffer.Len())
-	r.outBuffer.Reset()
+	line, err := l.buffer.CollectUntil(l.nl)
 	if err != nil {
-		// This should never happen as otherwise we have a broken state
-		panic(err)
+		return 0, err
 	}
-
-	// return and reset consumed bytes count
-	sz := r.byteCount
-	r.byteCount = 0
-	return bytes, sz, nil
+	return copy(buf, line), nil
 }
 
-// Reads from the buffer until a new line character is detected
-// Returns an error otherwise
-func (r *Reader) advance() error {
-	// Initial check if buffer has already a newLine character
-	idx := r.inBuffer.IndexFrom(r.inOffset, r.nl)
-
-	// fill inBuffer until '\n' sequence has been found in input buffer
-	for idx == -1 {
-		// increase search offset to reduce iterations on buffer when looping
-		newOffset := r.inBuffer.Len() - len(r.nl)
-		if newOffset > r.inOffset {
-			r.inOffset = newOffset
-		}
-
-		buf := make([]byte, r.bufferSize)
-
-		// try to read more bytes into buffer
-		n, err := r.reader.Read(buf)
-
-		// Appends buffer also in case of err
-		r.inBuffer.Append(buf[:n])
-		if err != nil {
-			return err
-		}
+func isNewLine(idx int) bool {
+	return idx != -1
+}
 
-		// empty read => return buffer error (more bytes required error)
-		if n == 0 {
-			return streambuf.ErrNoMoreBytes
-		}
+type decoderScanner struct {
+	reader     io.Reader
+	decoder    transform.Transformer
+	bufferSize int
+	byteCount  int
+}
 
-		// Check if buffer has newLine character
-		idx = r.inBuffer.IndexFrom(r.inOffset, r.nl)
+func newDecoderScanner(reader io.Reader, codec encoding.Encoding, bufferSize int) *decoderScanner {
+	return &decoderScanner{
+		reader:     reader,
+		decoder:    codec.NewDecoder(),
+		bufferSize: bufferSize,
+		byteCount:  0,
 	}
+}
 
-	// found encoded byte sequence for '\n' in buffer
-	// -> decode input sequence into outBuffer
-	sz, err := r.decode(idx + len(r.nl))
+func (d *decoderScanner) Scan() ([]byte, int, error) {
+	buf := make([]byte, d.bufferSize)
+	n, err := d.reader.Read(buf)
 	if err != nil {
-		logp.Err("Error decoding line: %s", err)
-		// In case of error increase size by unencoded length
-		sz = idx + len(r.nl)
+		return nil, 0, err
 	}
 
-	// consume transformed bytes from input buffer
-	err = r.inBuffer.Advance(sz)
-	r.inBuffer.Reset()
-
-	// continue scanning input buffer from last position + 1
-	r.inOffset = idx + 1 - sz
-	if r.inOffset < 0 {
-		// fix inOffset if '\n' has encoding > 8bits + firl line has been decoded
-		r.inOffset = 0
-	}
-
-	return err
+	return transform.Bytes(d.decoder, buf[:n])
```

Comment:
The true scanner is in `d.reader`. This method assume the underlying reader did already decode and 'split' the contents correctly. That is, the `lineReader` is still the scanner and the `decoderScanner` lifts the `lineReader` into a scanner + byte transformation.